### ISSUE TITLE
bemenu is displayed only in one monitor

### ISSUE THREAD

**Issue by zwma** [Cloudef/bemenu#102]
Building from https://github.com/Cloudef/bemenu/commit/cd53b7bb555cf1c5afaae3779a88e126571faf8c, Wayland backend.

Now bemenu only appears in display marked as 0 by default. (In my case it's an external monitor). I can change it to 1 using the option `-m`, but there is no way (that I know) to have it show up in both displays. It used to be like that in 0.3.0

**Comment by Cloudef** [comment]
https://github.com/Cloudef/bemenu/pull/103 Can you try if this PR works for you. You need to pass `-1` to `-m` option. This works only for wayland.

**Comment by dwma** [comment]
Just tried it and it didn't work, adding `-m -1` behaves like not having `-m` set at all.

`0` and `1` works as before although this time is reversed (don't know if the numbers are assigned randomly or the change is related to this PR)

FWIW, the compiler complained about comparing `uint` and `int`.

**Comment by Cloudef** [comment]
I think I messed up the if condition. I don't really have proper wayland setup here to test myself. Anyhow updated the PR, you can try again. As extra `all` option for `-m` should be equivalent to `-1` now too.

**Comment by dwma** [comment]
No complaints from the compiler this time but the end result is exactly the same :( I tried both `-1` and `all`.

**Comment by Cloudef** [comment]
Edited the pr again :)

**Comment by dwma** [comment]
No luck, same as before.

**Comment by Cloudef** [comment]
Dunno, the latest changes should work by only reading it. But since I don't have wayland setup I can't really test either. The PR could be worked on by someone else in the meantime.

**Comment by Cloudef** [comment]
Ah, I did notice one error still at my edits. Updated again.

**Comment by dwma** [comment]
Yep, worked this time!

**Comment by Cloudef** [comment]
Did I break anything else?

**Comment by dwma** [comment]
Tested the `-m` option with `0`, `1`, `-1` and `all`  and it's working fine.

However, now you can't keep pressed a key (backspace to delete, for example). This worked in https://github.com/Cloudef/bemenu/commit/cd53b7bb555cf1c5afaae3779a88e126571faf8c

**Comment by Cloudef** [comment]
Do you mean key repeat?

**Comment by dwma** [comment]
Yes, exactly.

**Comment by Cloudef** [comment]
For key repeat, I'm not aware of any recent commits that would've changed the behavior for wayland. If you can you should do a git bisect to find the commit that broke it.

**Comment by dwma** [comment]
Key repeat doesn't work only with the option `-m all/-1` enabled. Otherwise it works as expected.

**Comment by Cloudef** [comment]
https://github.com/Cloudef/bemenu/issues/98#issuecomment-599006997
can you try modifying the code as instructed here?

**Comment by dwma** [comment]
I changed the two lines to
```
    struct epoll_event ep[1];
    uint32_t num = epoll_wait(efd, ep, 1, -1);
```
but it didn't have any effect.

**Comment by Cloudef** [comment]
I see, not sure why the window count would affect the key repeat.

**Comment by dwma** [comment]
Is it feasible to show the menu only on the screen that has focus, thus avoiding this issue? Since that is ultimately what is achieved by showing it on all screens at once.

**Comment by Cloudef** [comment]
https://github.com/swaywm/wlr-protocols/blob/master/unstable/wlr-layer-shell-unstable-v1.xml#L50-L52 It seems to be possible.

https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/window.c#L327 if you replace the `output` on this line with `NULL`, it should behave like that.

**Comment by dwma** [comment]
It worked!
I also had to add ```allow-null="true"``` in https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/wlr-layer-shell-unstable-v1.xml#L55

**Comment by Cloudef** [comment]
The wayland protocol file hasn't been updated in a while. It would be nice to encode for example `-2` and `focused` to focused output for the `-m` option. I hope somebody can chime in and figure out the key repeat and other input issues under wayland though.

**Comment by Soundtoxin** [comment]
> I'm not aware of any recent commits that would've changed the behavior for wayland.

I can confirm that on version 0.3.0 from my distro's repo I don't have working key repeat either, running Sway.

**Comment by hkupty** [comment]
> Is it feasible to show the menu only on the screen that has focus? Since that is ultimately what is achieved by showing it on all screens at once.
I'm trying to achieve that by running: `swaymsg -r -t get_outputs | jq '.[] | select(.focused == true) | .id'`, but the `id` propertie does not match the monitor number you're expecting. I assume the values are provided by wlroots, which I also assume is what you're using under the hood for wayland, so shouldn't the id's match there?

**Comment by Cloudef** [comment]
Outputs actually have no id or such. The number is the nth output the compositor gives to bemenu. This means the behavior of the `-m` option depends on the underlying wayland compositor.

**Comment by hkupty** [comment]
So should I raise this to sway so I get the output/monitor index that is used by bemenu from `swaymsg -t get_outputs`? Or do you think that information is available elsewhere?

**Comment by Cloudef** [comment]
I think if you count the outputs in `get_outputs` and use the `nth` of particular output it should work. Unless the order of `get_outputs` doesn't match the order they come into bemenu from the compositor. https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/registry.c#L287

**Comment by Cloudef** [comment]
I just realized there is "id" here https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/registry.c#L287 this might be what `get_outputs` gives.

**Comment by hkupty** [comment]
Yes, I think it's correct. However the ID does not correspond to the index.
I checked by using the output order. It turns out it came in the opposite order this time, not sure if it will remain true after I reboot.

The command I used to run was the following:
`bemenu-run -m$(swaymsg -r -t get_outputs | jq '. | reverse | to_entries | .[] | select(.value.focused == true) | .key')`
And it correctly opened bemenu-run in the window I am focused. Just a disclaimer, I have two external monitors + the notebook display. The monitor at `0x0` is Index 0, but it's the last one of the list, which I believe to be accidental.

I assume indexes are being given according to distance to `0x0`, which would explain the embedded display being the last one.

However, I'm not really sure about what would be the *correct* oneliner that takes distance from `0x0` in account, if that is the actual parameter to be used.

**Comment by Cloudef** [comment]
If the "id" can be used to choose monitor more reliably, might make sense to change the bemenu to use the "id" instead. I have to check that it's actually unique id of output object and not just some interface identifier. There is also actually "focused" output support on the protocol itself, but not exposed as feature in bemenu yet.

**Comment by hkupty** [comment]
Yes, I can see two scenarios:
 - Picking the current screen;
 - Picking specifically one screen all the time (in my case, for example, it could be the laptop display, as it's almost always empty).

**Comment by shibumi** [comment]
> The command I used to run was the following:
> `bemenu-run -m$(swaymsg -r -t get_outputs | jq '. | reverse | to_entries | .[] | select(.value.focused == true) | .key')`
> And it correctly opened bemenu-run in the window I am focused. Just a disclaimer, I have two external monitors + the notebook display. The monitor at `0x0` is Index 0, but it's the last one of the list, which I believe to be accidental.

Yes, this seems accidentical. On my setup your command outputs a "0", but what I need is a "1". :S I tried removing the reversing, but this doesn't help. Any idea? Did you manage to get a stable script?

**Comment by shibumi** [comment]
Hi,
I found a stable solution for this issue. I use this script for now. It retrieves the focused output name (for example `HDMI-A-3`) and extracts the last number in the name via regex. It looks like the last number in the name is the index number we are looking for.

I am running this script via: `bemenu-run -m "$(script)"`. This works fine for me.

```python
#!/usr/bin/env python

import asyncio
from i3ipc.aio import Connection
import re


async def main():
    i3 = await Connection().connect()
    outputs = await i3.get_outputs()
    for output in outputs:
        if output.focused:
            return output.name

focus = asyncio.run(main())
match = re.search('.*-(\d)', focus)
print(match.group(1))
```

**Comment by farwayer** [comment]
Without python script:
```js
bemenu -m $(swaymsg -r -t get_outputs | jq '. [] | select (.focused == true) | .name | split ("-") | last')
```

**Comment by Cloudef** [comment]
This might be very sway specific behaviour. I think it might be safer if I used the id from the wayland output callback instead.

**Comment by shibumi** [comment]
@farwayer thanks alot. This is indeed easier.

**Comment by farwayer** [comment]
@Cloudef It will be nice to add flag to show bemenu on active output.

**Comment by msrd0** [comment]
@farwayer Thanks for sharing but your solution doesn't work for me. I have one monitor connected via HDMI (`HDMI-A-1`) and one via DVI (`DVI-D-1`), and that gives me `"1"` as the monitor no matter which one is selected. Seems like your script is very specific to your individual monitor setup.

**Comment by farwayer** [comment]
@msrd0 my snippet is just simple replacement for python code above. It works for me (HDMI-A-1, HDMI-A-3). To be honest I don't know what we should pass to `-m` flag. I tried `HDMI-A-1` but without success. Let's wait maybe @Cloudef change behaviour (it is expected that it should be displayed on the active monitor). It will be more correct to fix this issue inside code instead of complex hacks.

**Comment by shibumi** [comment]
One update. I moved to a new laptop and two external displays with Displayport.
Now my python script stopped working :( it only spawns on the same monitor.

**Comment by fancycade** [comment]
I've been able to test this out on my end and even recreate the listed bugs. A few notes,

1. After logging the id values in registry_handle_global, I'm pretty sure these are interface ids. It went from 1 all the way to 41 when I tested it out. So I don't think that is the answer.

2. I wrote a patch that would pass NULL to bm_wl_display_create when monitor is set to -2. As reported above this does mostly solve the problem. However, this led to a new bug, where the scale does not seem to be set correctly in that it comes out a bit pixelated. It seems that output->scale is not set correctly in recreate_windows. Interestingly, this issue does not appear when monitor is set -1 (all).

3. I'm able to recreate the inability to repeat keys when monitor is set to all. Even with latest master branch, I think this should be filed as a separate issue.

In summary, I think fixing the scaling issue I mentioned in 2 is the best bet.

**Comment by Cloudef** [comment]
@fancycade https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/registry.c#L295 can you log the id here, and check if the ids are not same for different outputs and stays consistent across runs? If that isn't true what we could do is assign id to the outputs ourselves by adding some logic in this function, https://github.com/Cloudef/bemenu/blob/master/lib/renderers/wayland/registry.c#L241 The top leftmost would always be output id 0, and the next 1, and so on ...

**Comment by Cloudef** [comment]
Not sure if we can rely on the id actually, assigning our own id based on the "physical location" of the monitor might be the best idea.

**Comment by fancycade** [comment]
@Cloudef 
```
Initial run with plugged in
ID: 36
ID: 37

Without plugged in
ID: 36

Plugged back in
ID: 36
ID: 38

Without plugged in
ID: 36
```

So I'm guessing my laptop is 36, and the HDMI I connect with is the 37 or 38. Note, I only have one HDMI device to test with.

Assigning id's seems like an option.

**Comment by Cloudef** [comment]
Seems like relying on any of the wayland id's is not good idea then. Assigning IDs ourselves based on the physical location seems like the best idea.

**Comment by shibumi** [comment]
> Seems like relying on any of the wayland id's is not good idea then. Assigning IDs ourselves based on the physical location seems like the best idea.

I don't know much about all of this, but shouldn't wayland have a deterministic way to calculate these IDs? So we should just look this up and then go the same way, I guess. Also I thought about the issue is about getting bemenu on the current focused monitor :D

**Comment by Cloudef** [comment]
> > Seems like relying on any of the wayland id's is not good idea then. Assigning IDs ourselves based on the physical location seems like the best idea.
> 
> I don't know much about all of this, but shouldn't wayland have a deterministic way to calculate these IDs? So we should just look this up and then go the same way, I guess. Also I thought about the issue is about getting bemenu on the current focused monitor :D

Default wayland protocol only exposes model, make, "x/y" and width/height of the output. There might be way to use some sort of "output configuration" wayland protocol to refer output with a name such as HDMI-1 or something. The best way if not using such protocol, would be to construct the ids ourselves, based on the "x/y" reported by default wayland protocol. That should keep the id consistent across runs. And yeah I guess this issue has been used to discuss the wayland output problems in general now.

**Comment by fancycade** [comment]
I wrote some code to assign id's to the outputs here https://github.com/fancycade/bemenu/commit/ce9bc6ae8e64babb9bde236c3ea9e20d63cec57f

I find it works a bit more intuitively in that my laptop is assigned id 0 and the monitor is 1. Whereas the current master branch is the opposite. Tested after a reboot and it still worked. Would be good to get other's to test this out.

With this we might be able to to implement the focus part, but we first we have to figure out which output is actually being focused.

**Comment by Cloudef** [comment]
>With this we might be able to to implement the focus part, but we first we have to figure out which output is actually being focused.

This can already be implemented as described here https://github.com/Cloudef/bemenu/issues/102#issuecomment-604462293

**Comment by fancycade** [comment]
Yes, but it wont properly have scaling in the case where the monitor/screens have different resolutions https://github.com/Cloudef/bemenu/issues/102#issuecomment-709472376, see point 2.

Still have to figure out which output is being used so as not to overwrite its scale with the other output scale.

**Comment by Cloudef** [comment]
This sounds like a limitation / design bug in the protocol indeed. I think the assigned output should be able to be queried or delivered to the client.

**Comment by timsofteng** [comment]
Hey. Have any news? I've catch this bug as well.

**Comment by fancycade** [comment]
I think the original issue has been resolved and this can be closed.

Separate issues have been created for the other problems that have been identified.

**Comment by shibumi** [comment]
For me the issue has not been solved. It has been solved for a short time and right now I have the same behavior again (this might be a problem in my script...).

What is the current suggested way to solve this?

**Comment by fancycade** [comment]
@shibumi Depends on what you want to do.

If you want to have bemenu on a particular monitor every time then you can specify the string name of the argument:

```
bemenu-run -m HDMI-3
```

After #174, if you just want the focused monitor then running bemenu-run with no arguments will work fine. It leaves it up to the compositor to choose by default, same behavior as:

```
bemenu-run -m focused
```

**Comment by shibumi** [comment]
@fancycade  thanks! This means I could finally delete my custom scripts for it that didn't work properly.

**Comment by tuxinaut** [comment]
@fancycade Thx :smiley:
